lib/posix-socket: Add UK_LIBC_SYSCALLS guard around send() and recv()

razvand opened the pull request:
`send()` and `recv()` are defined as wrappers in `posix-socket` to be used in case they aren't already defined by a libc (such as Musl). Once a libc is used, there will be a conflict, since the libc would define the `send()` and `recv()` functions itself.

This commits adds a `#if UK_LIBC_SYSCALLS` guard around the two functions, such that they will be defined only when a libc is not present. If a libc (such as Musl) is present, the functions are masked out.
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]

### Additional configuration

Test this PR using Musl. You can use [these scripts](https://github.com/unikraft-upb/scripts/tree/main/musl).

razvand committed:
lib/posix-socket: Add UK_LIBC_SYSCALLS guard around send() and recv()

`send()` and `recv()` are defined as wrappers in `posix-socket` to be
used in case they aren't already defined by a libc (such as Musl). Once
a libc is used, there will be a conflict, since the libc would define
the `send()` and `recv()` functions itself.

This commits adds a `#if UK_LIBC_SYSCALLS` guard around the two
functions, such that they will be defined only when a libc is not
present. If a libc (such as Musl) is present, the functions are masked
out.

Signed-off-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>

StefanJum reviewed (commented):
@razvand `send` and `recv` do not appear to be [system calls](https://filippo.io/linux-syscall-table/).
If we add a dependency on `UK_LIBC_SYSCALLS`, they will not be available unless `syscall_shim` is enabled, which I don't know if it's the right approach considering they are not syscalls.

Does it make more sense to move `send` and `recv` to `nolibc`, since most likely any `libc` will provide their own implementation (at least `newlib` and `musl` both do)?

If not, we can also add a new `provide send and recv wrappers` config option to `posix-socket` so we can specify if we want them defined or not.
In any way, I don't think they should depend on `syscall_shim`.

razvand commented:
>If we add a dependency on UK_LIBC_SYSCALLS, they will not be available unless syscall_shim is enabled, which I don't know if it's the right approach considering they are not syscalls.

But, @StefanJum , their implementation depends on `sendto` or `recvfrom` that are system calls. You still need to include `posix-socket` and you still need to include `uk/syscall.h` (included by default). If the `syscall_shim` is NOT enabled the default value of `UK_LIBC_SYSCALLS` is `1` and the two functions (`send()` and `recv()`) are enabled. See [here](https://github.com/unikraft/unikraft/blob/staging/lib/syscall_shim/include/uk/syscall.h#L50). The only way to disable then is to enable the `syscall_shim` and the enable the `CONFIG_LIBSYSCALL_SHIM_NOWRAPPER` configuration option.

StefanJum commented:
> But, @StefanJum , their implementation depends on `sendto` or `recvfrom` that are system calls. You still need to include `posix-socket` and you still need to include `uk/syscall.h` (included by default). If the `syscall_shim` is NOT enabled the default value of `UK_LIBC_SYSCALLS` is `1` and the two functions (`send()` and `recv()`) are disabled. See [here](https://github.com/unikraft/unikraft/blob/staging/lib/syscall_shim/include/uk/syscall.h#L50). The only way to disable then is to enable the `syscall_shim` and the enable the `CONFIG_LIBSYSCALL_SHIM_NOWRAPPER` configuration option.

Ah that makes sense.
Another question I have is do we want to use `recvfrom` and `sendto` wrappers or should we change to using our syscalls (`uk_syscall_r_recvfrom` and `uk_syscall_r_sendto`)? This is more related to @mariasfiraiala s pr on `vsfcore` #636.

It should not make a difference here, since `send` and `recv` will not be enabled unless `UK_LIBC_SYSCALLS` is enabled, and then the `recvfrom` and `sendto` wrappers will just call our syscalls, it's just a note for consistency.

Other than that it looks good, I will review it tomorrow after some more testing.

StefanJum commented:
> Ah that makes sense. Another question I have is do we want to use `recvfrom` and `sendto` wrappers or should we change to using our syscalls (`uk_syscall_r_recvfrom` and `uk_syscall_r_sendto`)? This is more related to @mariasfiraiala s pr on `vsfcore` #636.
> 
> It should not make a difference here, since `send` and `recv` will not be enabled unless `UK_LIBC_SYSCALLS` is enabled, and then the `recvfrom` and `sendto` wrappers will just call our syscalls, it's just a note for consistency.

@razvand  I tested this and it works fine, it's just the comment above now.

razvand commented:
> @razvand I tested this and it works fine, it's just the comment above now.

@StefanJum , I'll address your comments today.

razvanvirtan commented on lib/posix-socket/socket.c:
```diff
@@ -463,11 +463,13 @@ UK_SYSCALL_R_DEFINE(ssize_t, recvfrom, int, sock, void *, buf, size_t, len,
 
-/* Not a system call */
+#if UK_LIBC_SYSCALLS
+/* Provide wrapper if not provided by Musl) or some other libc. */
```
I think the `)` after Musl doesn't have any purpose here.

razvanvirtan commented on lib/posix-socket/socket.c:
```diff
@@ -576,11 +578,13 @@ UK_SYSCALL_R_DEFINE(ssize_t, sendto, int, sock, const void *, buf, size_t, len,
 
-/* Not a system call */
+#if UK_LIBC_SYSCALLS
+/* Provide wrapper if not provided by Musl) or some other libc. */
```
Same comment about the `)` as above.

razvanvirtan reviewed (commented):
Thanks, @razvand!
I only have to comments on typos. Other than that, please also address the comments from @StefanJum and then I think we are safe to merge.

razvand committed:
lib/posix-socket: Use uk_syscall_e versions for recvfrom / sendto

This is not strictly required. But it is for consistency with other
functions in internal libraries, that also use uk_syscall_r or
uk_syscall_e variants. Those variants are useful to ensure calls to
local implementation. The libc-variants (without the uk_syscall_r /
uk_syscall_e) may clash with libc symbols.

Signed-off-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>

razvand commented:
@StefanJum , @razvanvirtan , I addressed your comments.

StefanJum reviewed (approved):
All good on my side.
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

mariasfiraiala reviewed (approved):
Looks good. Thanks @razvand !

Reviewed-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

razvanvirtan reviewed (approved):
Everything looks good here. Thank you @razvand !

Approved-by: Razvan Virtan <virtanrazvan@gmail.com>
